<?xml version="1.0"?>
<robot name="lophorina">

  <material name="gray">
    <color rgba="0.8 0.8 0.8 1"/>
  </material>

  <material name="blue">
    <color rgba="0 0 1 1"/>
  </material>

  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>

  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>

  <material name="green">
    <color rgba="0 1 0 1"/>
  </material>

  <link name="odom">  
  </link>  

  <joint name="odom_to_base_link" type="fixed">
    <parent link="odom"/>
    <child link="base_link"/>
  </joint>

  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/robot_body.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin rpy="0 0 1.57079632679"/>
      <material name="black"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.05"/>
      <geometry>
        <box size="0.6 0.8 0.1"/>        
      </geometry>
    </collision>
    <inertial>
	  <!-- Volume:  1.2327948124e+06 -->
	  <mass value="1.5"/>

	  <!-- Center of mass:  1.4152750000e-02  7.9666542000e-01  2.5105856930e+01 -->
	  <pose>  1.4166e-05  7.96665e-04  2.5105858e-02 0 0 0 </pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="4.1954e-02"
	    ixy="1.6879e-08"
	    ixz="2.2501e-09"
	    iyy="6.8166e-03"
	    iyz="-2.2635e-06"
	    izz="4.8090e-02"/>
	  
   </inertial>
  </link>
  <link name="motor_front_right">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/tilt_motor.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="blue"/>
      <origin rpy="0 0 1.57079632679" xyz="0 0 0"/>
    </visual>
    <inertial>
	  <!-- Volume:  4.3863842816e+04 -->
	  <mass value="0.15"/>

	  <!-- Center of mass:  1.1607480469e+02  1.2213770750e+01  1.8455069580e+01 -->
	  <pose>  1.16074806e-01 1.2213773e-02 1.8455067e-02 0 0 0 </pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="6.5200e-05"
	    ixy="-1.8405e-06"
	    ixz="-4.4279e-05"
	    iyy="2.7136e-04"
	    iyz="-5.5407e-07"
	    izz="2.2969e-04"/>
	  
    </inertial>
  </link>

  <link name="motor_front_left">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/tilt_motor.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="blue"/>
      <origin rpy="0 0 -1.57079632679" xyz="0 0 0"/>
    </visual>
   <inertial>
	  <!-- Volume:  4.3863842816e+04 -->
	  <mass value="0.15"/>

	  <!-- Center of mass:  1.1607480469e+02  1.2213770750e+01  1.8455069580e+01 -->
	  <pose>  1.16074806e-01 1.2213773e-02 1.8455067e-02 0 0 0 </pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="6.5200e-05"
	    ixy="-1.8405e-06"
	    ixz="-4.4279e-05"
	    iyy="2.7136e-04"
	    iyz="-5.5407e-07"
	    izz="2.2969e-04"/>
	
   </inertial>
  </link>

  <link name="motor_back_right">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/tilt_motor.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="gray"/>
      <origin rpy="0 0 1.57079632679" xyz="0 0 0"/>
    </visual>
    <inertial>
	  <!-- Volume:  4.3863842816e+04 -->
	  <mass value="0.15"/>

	  <!-- Center of mass:  1.1607480469e+02  1.2213770750e+01  1.8455069580e+01 -->
	  <pose>  1.16074806e-01 1.2213773e-02 1.8455067e-02 0 0 0 </pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="6.5200e-05"
	    ixy="-1.8405e-06"
	    ixz="-4.4279e-05"
	    iyy="2.7136e-04"
	    iyz="-5.5407e-07"
	    izz="2.2969e-04"/>
	  
</inertial>
  </link>

  <link name="motor_back_left">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/tilt_motor.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="gray"/>
      <origin rpy="0 0 -1.57079632679" xyz="0 0 0"/>
    </visual>
    <inertial>
	  <!-- Volume:  4.3863842816e+04 -->
	  <mass value="0.15"/>

	  <!-- Center of mass:  1.1607480469e+02  1.2213770750e+01  1.8455069580e+01 -->
	  <pose>  1.16074806e-01 1.2213773e-02 1.8455067e-02 0 0 0 </pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="6.5200e-05"
	    ixy="-1.8405e-06"
	    ixz="-4.4279e-05"
	    iyy="2.7136e-04"
	    iyz="-5.5407e-07"
	    izz="2.2969e-04"/>
	  
   </inertial>
  </link>

  <link name="front_right_propeller">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/cw_propeller.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0"/>
      <material name="green"/>
    </visual>
    <inertial>
	  <!-- Volume:  6.4020546560e+03 -->
	  <mass value="0.008"/>

	  <!-- Center of mass: -7.6309433000e-01 -1.5900170000e-02  5.6214502000e+00 -->
	  <pose> 0 0 5.621359e-03 0 0 0</pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="1.0083e-06"
	    ixy="-4.9698e-06"
	    ixz="-1.5179e-08"
	    iyy="3.3832e-05"
	    iyz="-1.8552e-09"
	    izz="3.4773e-05"/>
	  
   </inertial>   
  </link>
  <gazebo reference="front_right_propeller">
    <material>Gazebo/Orange</material>
  </gazebo>

  <gazebo>
    <plugin name="front_right_propeller_model_push" filename="libhelice_push.so">
          <link_name>front_right_propeller</link_name>
          <force>20</force>
    </plugin>
  </gazebo>


  <link name="front_left_propeller">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/ccw_propeller.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
	     <!-- Volume:  6.4020546560e+03 -->
	     <mass value="0.008"/>

	     <!-- Center of mass: -1.2450975000e+00  4.7389927000e-01  5.6214502000e+00 -->
	     <pose> 0 0 5.621359e-03 0 0 0</pose>

	     <!-- Inertia matrix -->
	     <inertia
	 	    ixx="1.0084e-06"
	    	ixy="4.9698e-06"
	    	ixz="-1.5181e-08"
	    	iyy="3.3832e-05"
	    	iyz="1.8550e-09"
	    	izz="3.4773e-05"/>
	     
   </inertial>
  </link>
  <gazebo>
    <plugin name="front_left_propeller_model_push" filename="libhelice_push.so">
          <link_name>front_left_propeller</link_name>
          <force>20</force>
          <dir>ccw</dir>
    </plugin>
  </gazebo>

  <gazebo reference="front_left_propeller">
    <material>Gazebo/Red</material>
  </gazebo>


  <link name="back_left_propeller">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/cw_propeller.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="green"/>
    </visual>
    <inertial>
	  <!-- Volume:  6.4020546560e+03 -->
	  <mass value="0.008"/>

	  <!-- Center of mass: -7.6309433000e-01 -1.5900170000e-02  5.6214502000e+00 -->
	  <pose> 0 0 5.621359e-03 0 0 0</pose>

	  <!-- Inertia matrix -->
	  <inertia
	    ixx="1.0083e-06"
	    ixy="-4.9698e-06"
	    ixz="-1.5179e-08"
	    iyy="3.3832e-05"
	    iyz="-1.8552e-09"
	    izz="3.4773e-05"/>
	  
   </inertial>
  </link>

  <gazebo reference="back_left_propeller">
    <material>Gazebo/Green</material>
  </gazebo>

  <gazebo>
    <plugin name="back_left_propeller_model_push" filename="libhelice_push.so">
          <link_name>back_left_propeller</link_name>
          <force>20</force>
    </plugin>
  </gazebo>

  <link name="back_right_propeller">
    <visual>
      <geometry>
        <mesh filename="package://lophorina/meshes/ccw_propeller.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
	    <!-- Volume:  6.4020546560e+03 -->
	    <mass value="0.008"/>

      <!-- Center of mass: -1.2450975000e+00  4.7389927000e-01  5.6214502000e+00 -->
      <pose> 0 0 5.621359e-03 0 0 0</pose>

      <!-- Inertia matrix -->
      <inertia
	 	    ixx="1.0084e-06"
	    	ixy="4.9698e-06"
	    	ixz="-1.5181e-08"
	    	iyy="3.3832e-05"
	    	iyz="1.8550e-09"
	    	izz="3.4773e-05"/>

   </inertial>
  </link>

  <gazebo reference="back_right_propeller">
    <material>Gazebo/Blue</material>
  </gazebo>

  <gazebo>
    <plugin name="back_right_propeller_model_push" filename="libhelice_push.so">
          <link_name>back_right_propeller</link_name>
          <force>20</force>
          <dir>ccw</dir>
    </plugin>
  </gazebo>
  
  <joint name="base_link_to_motor_front_right" type="revolute">
    <limit effort="1000.0" velocity="0.5" lower="-0.78539816339" upper="0.78539816339"/>
    <parent link="base_link"/>
    <child link="motor_front_right"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0.1825 0.126 0.025"/>
  </joint>

  <transmission name="base_link_to_motor_front_right_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$base_link_to_motor_front_right_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="base_link_to_motor_front_right">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="base_link_to_motor_front_left" type="revolute">
    <limit effort="1000.0" velocity="0.5" lower="-0.78539816339" upper="0.78539816339"/>
    <parent link="base_link"/>
    <child link="motor_front_left"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0.1825 -0.126 0.025"/>
  </joint>

  <transmission name="base_link_to_motor_front_left_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$base_link_to_motor_front_left_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="base_link_to_motor_front_left">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="base_link_to_motor_back_right" type="revolute">
    <limit effort="1000.0" velocity="0.5" lower="-0.78539816339" upper="0.78539816339"/>
    <parent link="base_link"/>
    <child link="motor_back_right"/>
    <axis xyz="0 1 0"/>
    <origin xyz="-0.1825 0.126 0.025"/>
  </joint>

  <transmission name="base_link_to_motor_back_right_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$base_link_to_motor_back_right_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="base_link_to_motor_back_right">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="base_link_to_motor_back_left" type="revolute">
    <limit effort="1000.0" velocity="0.5" lower="-0.78539816339" upper="0.78539816339"/>
    <parent link="base_link"/>
    <child link="motor_back_left"/>
    <axis xyz="0 1 0"/>
    <origin xyz="-0.1825 -0.126 0.025"/>
  </joint>

  <transmission name="base_link_to_motor_back_left_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$base_link_to_motor_back_left_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="base_link_to_motor_back_left">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="motor_front_right_to_front_right_propeller" type="continuous">
    <limit effort="100" velocity="10000"/>
    <parent link="motor_front_right"/>
    <child link="front_right_propeller"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 0.138 0.05"/>
  </joint>

  <transmission name="motor_front_right_to_front_right_propeller_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$motor_front_right_to_front_right_propeller_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="motor_front_right_to_front_right_propeller">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="motor_front_left_to_front_left_propeller" type="continuous">
    <limit effort="100" velocity="10000"/>
    <parent link="motor_front_left"/>
    <child link="front_left_propeller"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 -0.138 0.05"/>
  </joint>

  <transmission name="motor_front_left_to_front_left_propeller_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$motor_front_left_to_front_left_propeller_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="motor_front_left_to_front_left_propeller">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="motor_back_right_to_back_right_propeller" type="continuous">
    <limit effort="100" velocity="10000"/>
    <parent link="motor_back_right"/>
    <child link="back_right_propeller"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 0.138 0.05"/>
  </joint>

  <transmission name="motor_back_right_to_back_right_propeller_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$motor_back_right_to_back_right_propeller_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="motor_back_right_to_back_right_propeller">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <joint name="motor_back_left_to_back_left_propeller" type="continuous">
    <limit effort="100" velocity="10000"/>
    <parent link="motor_back_left"/>
    <child link="back_left_propeller"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0 -0.138 0.05"/>
  </joint>

  <transmission name="motor_back_left_to_back_left_propeller_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$motor_back_left_to_back_left_propeller_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="motor_back_left_to_back_left_propeller">
      <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <!-- Gazebo plugin for ROS Control -->
  <gazebo>
    <plugin name="lopho_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/lophorina</robotNamespace>
    </plugin>
  </gazebo>

</robot>
